### ISSUE TITLE
IllegalStateException: Cannot register not yet registered manager

### ISSUE THREAD

**Issue by electrolobzik** [bumptech/glide#1545]
**Glide Version**: 4.0.0-SNAPSHOT

**Integration libraries**: okhttp-integration:2.0.0-SNAPSHOT'

**Device/Android Version**: Android 4.4.2, Samsung Galaxy S4

**Issue details / Repro steps / Use case background**: 

I use `Glide` in fragment `A` than I open fragment `B` return to the `A` and open `C`. At the moment of second disappearing of `A` I am getting this error.
I clear requests in `RequestManager` in fragment's `onDestroyView` via `RequestManager::onDestroy`. Seems like `RequestManager` for `A` was got from existing `SupportRequestManagerFragment` in [RequestManagerRetriever:L219](https://github.com/bumptech/glide/blob/master/library/src/main/java/com/bumptech/glide/manager/RequestManagerRetriever.java#L219) and was not registered in [Glide:L504](https://github.com/bumptech/glide/blob/master/library/src/main/java/com/bumptech/glide/Glide.java#L504). But the same instance of `RequestManager` was previously already unregistered in the first `onDestroyView`. I suppose `RequestManager` should register in `else` section of [RequestManagerRetriever:L219](https://github.com/bumptech/glide/blob/master/library/src/main/java/com/bumptech/glide/manager/RequestManagerRetriever.java#L219) like it does in case when `RequestManager` is created first time in `if` section.

**Glide load line / `GlideModule` (if any) / list Adapter code (if any)**:
```java
Glide.with(fragment)
```

**Stack trace / LogCat**:
```
java.lang.IllegalStateException: Cannot register not yet registered manager
at com.bumptech.glide.Glide.unregisterRequestManager(Glide.java:516)
at com.bumptech.glide.RequestManager.onDestroy(RequestManager.java:276)
at com.fsof.ui.fragment.BaseFragment.onDestroyView(BaseFragment.java:91)
at com.fsof.ui.fragment.BaseMvpFragment.onDestroyView(BaseMvpFragment.java:118)
at com.fsof.ui.fragment.BaseMvpLceFragment.onDestroyView(BaseMvpLceFragment.java:98)
at android.support.v4.app.Fragment.performDestroyView(Fragment.java:2313)
at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1188)
at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1295)
at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1277)
at android.support.v4.app.FragmentManagerImpl.dispatchDestroyView(FragmentManager.java:2185)
at android.support.v4.app.Fragment.performDestroyView(Fragment.java:2309)
at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1188)
at android.support.v4.app.FragmentManagerImpl.removeFragment(FragmentManager.java:1392)
at android.support.v4.app.BackStackRecord.run(BackStackRecord.java:758)
at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1682)
at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManager.java:541)
at android.os.Handler.handleCallback(Handler.java:733)
at android.os.Handler.dispatchMessage(Handler.java:95)
at android.os.Looper.loop(Looper.java:157)
at android.app.ActivityThread.main(ActivityThread.java:5356)
at java.lang.reflect.Method.invokeNative(Native Method)
at java.lang.reflect.Method.invoke(Method.java:515)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1265)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1081)
at dalvik.system.NativeStart.main(Native Method)
```

**Comment by TWiStErRob** [comment]
It looks like the manual `onDestroy` call hack from #1337 doesn't work in Glide v4, because that version does a lot more in [onDestroy](https://github.com/bumptech/glide/blame/master/library/src/main/java/com/bumptech/glide/RequestManager.java#L272) than the [previous version](https://github.com/bumptech/glide/blame/3.0/library/src/main/java/com/bumptech/glide/RequestManager.java#L221). In v3 it was just a way to call `requestTracker.clearRequests();`, but in v4 you may need to acquire the `requestTracker` via reflection; or create a PR for #1337.

**Comment by electrolobzik** [comment]
But what's the difference where am I calling `onDestroy` if it doesn't work second time? 

Btw, I will try to reproduce it later. Now I have to return to the `3.7.0`. =(

**Comment by TWiStErRob** [comment]
The difference is that `RequestManager.onDestroy` was designed (and assumed) to be called from `Fragment.onDestroy`, but you're calling earlier. Based on that assumption it's correct to unregister and expect the RequestManager to be dead after the call, because the fragment is dead too. You calling it earlier just confused Glide v4.

**Comment by electrolobzik** [comment]
Yes, seems that you are right. Thank you for the help. But it means that `Glide v4` is not ready for the production when using it in fragments with backstack. There is no simple way to avoid leaking of views inside fragments.

**Comment by TWiStErRob** [comment]
If that's your measure then v3 wasn't ready for production either, because `onDestroy` is not part of the public API of Glide, it's just a side effect of how Java interfaces work (their impl has to be `public`).

I think the best would be to fix the root cause of this issue, which is likely `Glide.clear` doesn't remove the request from the tracker, and/or fix #1337 properly.
